File: 445.两数相加II/main.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var result *ListNode
	s1, s2 := []int{}, []int{}
	cur := 0

	for l1 != nil {
		s1 = append(s1, l1.Val)
		l1 = l1.Next
	}
	for l2 != nil {
		s2 = append(s2, l2.Val)
		l2 = l2.Next
	}

	for len(s1) > 0 || len(s2) > 0 || cur != 0 {
		if len(s1) > 0 {
			cur += s1[len(s1)-1]
			s1 = s1[:len(s1)-1]
		}
		if len(s2) > 0 {
			cur += s2[len(s2)-1]
			s2 = s2[:len(s2)-1]
		}
		node := &ListNode{Val: cur % 10}
		cur /= 10
		node.Next = result
		result = node
	}

	return result
}

File: 445.两数相加II/main.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut s1 = Vec::with_capacity(128);
        let mut s2 = Vec::with_capacity(128);
        let mut cur = 0;

        let mut tmp1 = l1;
        while let Some(b) = tmp1 {
            tmp1 = b.next;
            s1.push(b.val);
        }
        let mut tmp2 = l2;
        while let Some(b) = tmp2 {
            tmp2 = b.next;
            s2.push(b.val);
        }
    }
}
